File: backend/internal/shared/events/envelope_test.go
package events

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

func TestEnvelopeRoundTrip(t *testing.T) {
	fixtures := readEnvelopeFixtures(t)
	if got, want := len(fixtures), 3; got != want {
		t.Fatalf("fixture count = %d, want %d", got, want)
	}

	for _, fixture := range fixtures {
		encoded, err := json.Marshal(fixture)
		if err != nil {
			t.Fatalf("Marshal(%s): %v", fixture.EventName, err)
		}
		var decoded Envelope
		if err := json.Unmarshal(encoded, &decoded); err != nil {
			t.Fatalf("Unmarshal(%s): %v", fixture.EventName, err)
		}
		if decoded.EventID != fixture.EventID ||
			decoded.EventName != fixture.EventName ||
			decoded.EventVersion != fixture.EventVersion ||
			decoded.AggregateType != fixture.AggregateType ||
			decoded.AggregateID != fixture.AggregateID ||
			decoded.Producer != fixture.Producer ||
			!decoded.OccurredAt.Equal(fixture.OccurredAt) {
			t.Fatalf("round-trip mismatch for %s: got %#v want %#v", fixture.EventName, decoded, fixture)
		}
		if (decoded.TraceID == nil) != (fixture.TraceID == nil) {
			t.Fatalf("traceId nil mismatch for %s", fixture.EventName)
		}
		if decoded.TraceID != nil && *decoded.TraceID != *fixture.TraceID {
			t.Fatalf("traceId mismatch for %s: got %q want %q", fixture.EventName, *decoded.TraceID, *fixture.TraceID)
		}
		if !jsonEqual(decoded.Payload, fixture.Payload) {
			t.Fatalf("payload mismatch for %s: got %s want %s", fixture.EventName, decoded.Payload, fixture.Payload)
		}
	}
}

func TestTraceIDSoftRequired(t *testing.T) {
	fixtures := readEnvelopeFixtures(t)
	missing := findEnvelope(t, fixtures, EventNameReportGenerated)
	warnings := missing.ValidateForPublish()
	if got, want := len(warnings), 1; got != want {
		t.Fatalf("missing traceId warnings = %d, want %d", got, want)
	}
	if warnings[0].Field != "traceId" || warnings[0].EventName != EventNameReportGenerated {
		t.Fatalf("unexpected warning: %#v", warnings[0])
	}

	present := findEnvelope(t, fixtures, EventNameTargetImportRequested)
	if present.TraceID == nil || *present.TraceID != "trace-target-import-1" {
		t.Fatalf("traceId not preserved: %#v", present.TraceID)
	}
	if warnings := present.ValidateForPublish(); len(warnings) != 0 {
		t.Fatalf("present traceId warnings = %#v, want none", warnings)
	}
}

func readEnvelopeFixtures(t *testing.T) []Envelope {
	t.Helper()
	wd, err := os.Getwd()
	if err != nil {
		t.Fatalf("Getwd: %v", err)
	}
	path := filepath.Join(wd, "..", "..", "..", "..", "shared", "events", "__fixtures__", "envelopes.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fixtures []Envelope
	if err := json.Unmarshal(data, &fixtures); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}
	return fixtures
}

func findEnvelope(t *testing.T, fixtures []Envelope, name EventName) Envelope {
	t.Helper()
	for _, fixture := range fixtures {
		if fixture.EventName == name {
			return fixture
		}
	}
	t.Fatalf("missing fixture for %s", name)
	return Envelope{}
}

func jsonEqual(left, right json.RawMessage) bool {
	var leftValue any
	var rightValue any
	if err := json.Unmarshal(left, &leftValue); err != nil {
		return false
	}
	if err := json.Unmarshal(right, &rightValue); err != nil {
		return false
	}
	leftBytes, _ := json.Marshal(leftValue)
	rightBytes, _ := json.Marshal(rightValue)
	return bytes.Equal(leftBytes, rightBytes)
}

File: frontend/src/lib/events/envelope.test.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { resolve } from 'node:path';
import { describe, expect, it } from 'vitest';

import {
  validateEnvelopeForPublish,
  type EventEnvelope,
} from './envelope';
import type {
  EventName,
  EventNameToPayload,
} from './events';

type FixtureEnvelope = EventEnvelope<EventNameToPayload[EventName]> & { eventName: EventName };

const fixturePath = resolve(fileURLToPath(new URL('../../../../shared/events/__fixtures__/envelopes.json', import.meta.url)));
const fixtures = JSON.parse(readFileSync(fixturePath, 'utf8')) as FixtureEnvelope[];

describe('event envelope contract', () => {
  it('round-trips shared fixture envelopes', () => {
    expect(fixtures).toHaveLength(2);

    for (const fixture of fixtures) {
      const decoded = JSON.parse(JSON.stringify(fixture)) as FixtureEnvelope;

      expect(decoded.eventId).toBe(fixture.eventId);
      expect(decoded.eventName).toBe(fixture.eventName);
      expect(decoded.eventVersion).toBe(fixture.eventVersion);
      expect(decoded.aggregateType).toBe(fixture.aggregateType);
      expect(decoded.aggregateId).toBe(fixture.aggregateId);
      expect(decoded.occurredAt).toBe(fixture.occurredAt);
      expect(decoded.producer).toBe(fixture.producer);
      expect(decoded.traceId).toBe(fixture.traceId);
      expect(decoded.payload).toEqual(fixture.payload);
    }
  });

  it('treats traceId as soft-required for publish', () => {
    const missing = fixtures.find((fixture) => fixture.eventName === 'report.generated');
    expect(missing).toBeDefined();
    expect(validateEnvelopeForPublish(missing!).map((warning) => warning.field)).toEqual(['traceId']);

    const present = fixtures.find((fixture) => fixture.eventName === 'target.import.requested');
    expect(present?.traceId).toBe('trace-target-import-1');
    expect(validateEnvelopeForPublish(present!)).toEqual([]);
  });
});
